File: python-restate-only/tour-of-agents/app/human_approval_agent_with_timeout.py
"""
Human-in-the-Loop Pattern

Implement resilient human approval steps that suspend execution until feedback is received.
Durable promises survive crashes and can be recovered across process restarts.
"""

from datetime import timedelta

import restate
from pydantic import BaseModel
from restate import RunOptions

from util.litellm_call import llm_call
from util.util import InsuranceClaim, request_review, tool


class ClaimPrompt(BaseModel):
    message: str = (
        "Process my hospital bill of 2024-10-01 for 3000USD for a broken leg at General Hospital."
    )


# TOOL IMPLEMENTATION
# <start_here>
async def request_human_approval(ctx: restate.Context, claim: InsuranceClaim) -> str:
    # Create a recoverable approval promise
    approval_id, approval_promise = ctx.awakeable(type_hint=str)

    await ctx.run_typed(
        "Request review", request_review, claim=claim, approval_id=approval_id
    )

    # Suspend until human resolves the approval or until the timeout hits
    # Check the service logs to see how to resolve it over HTTP, e.g.:
    # curl http://localhost:8080/restate/awakeables/sign_.../resolve --json '"approved"'
    match await restate.select(
        approval=approval_promise,
        timeout=ctx.sleep(timedelta(hours=3)),
    ):
        case ["approval", approved]:
            return approved
        case _:
            return "Approval timed out - Evaluate with AI"
# <end_here>


claim_approval_agent = restate.Service("HumanClaimApprovalAgent")


@claim_approval_agent.handler()
async def run(ctx: restate.Context, req: ClaimPrompt) -> str | None:
    """Evaluate an insurance claim with optional human approval for high-value claims."""

    # LLM evaluates the claim and decides if human approval is needed
    result = await ctx.run_typed(
        "Evaluate claim",
        llm_call,  # Use your preferred LLM SDK here
        RunOptions(max_attempts=3),
        messages=f"""You are an insurance claim evaluation agent. Use these rules:
        - if the amount is more than 1000, ask for human approval using tools;
        - if the amount is less than 1000, decide by yourself.
        Claim: {req.message}""",
        tools=[
            tool(
                "request_human_approval",
                "Ask for human approval for high-value claims",
                InsuranceClaim.model_json_schema(),
            )
        ],
    )

    # If the LLM requests human approval, suspend until a human resolves it
    if (
        result.tool_calls
        and result.tool_calls[0].function.name == "request_human_approval"
    ):
        claim = InsuranceClaim.model_validate_json(
            result.tool_calls[0].function.arguments
        )
        return await request_human_approval(ctx, claim)

    return result.content


if __name__ == "__main__":
    import asyncio
    import hypercorn

    app = restate.app(services=[claim_approval_agent])

    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))

File: typescript-restate-only/tour-of-agents/src/human-approval-agent-with-timeout.ts
/**
 * Human-in-the-Loop Pattern with Timeout
 *
 * Implement resilient human approval steps that suspend execution until feedback is received.
 * Durable promises survive crashes and can be recovered across process restarts.
 * Adds a timeout to the approval step to meet SLAs.
 */
import * as restate from "@restatedev/restate-sdk";
import { Context, TimeoutError } from "@restatedev/restate-sdk";
import { requestClaimReview, zodPrompt } from "./utils/utils";
import { tool } from "ai";
import z from "zod";
import llmCall from "./utils/llm";

const examplePrompt =
  "Process my hospital bill of 2024-10-01 for 3000USD for a broken leg at General Hospital.";

const tools = {
  humanApproval: tool({
    description: "Ask for human approval for high-value claims.",
    inputSchema: z.object({}),
  }),
};

// <start_here>
async function run(ctx: Context, { message }: { message: string }) {
  const prompt =
    "You are an insurance claim evaluation agent. Use these rules: " +
    "* if the amount is more than 1000, ask for human approval, " +
    "* if the amount is less than 1000, decide by yourself. " +
    `Evaluate this claim: ${message}`;
  const { text, toolCalls } = await ctx.run(
    "LLM call",
    // Use your preferred LLM SDK here
    async () => llmCall(prompt, tools),
    { maxRetryAttempts: 3 },
  );

  if (toolCalls?.[0]?.toolName === "humanApproval") {
    // Create a recoverable approval promise
    const approval = ctx.awakeable<boolean>();
    await ctx.run("request-review", () =>
      requestClaimReview(message, approval.id),
    );

    try {
      // At most 3 hours, to reach our SLA
      return await approval.promise.orTimeout({ hours: 3 });
    } catch (e) {
      if (e instanceof TimeoutError) {
        return {
          approved: false,
          reason: "Approval timed out - Evaluate with AI",
        };
      }
      throw e;
    }
  }

  return text;
}
// <end_here>

const agentService = restate.service({
  name: "HumanClaimApprovalWithTimeoutsAgent",
  handlers: {
    run: restate.createServiceHandler({ input: zodPrompt(examplePrompt) }, run),
  },
});

restate.serve({ services: [agentService], port: 9080 });
